### ISSUE TITLE
ilab model train produces Val loss nan with MLX on macOS (Merlinite-7b)

### ISSUE THREAD

**Issue by dgoldwx2112** [instructlab/instructlab#3287]
**Describe the bug**
When running "ilab model train --model-dir ./merlinite-7b-base --input-dir ./generated --iters 5 --local", training produces the following error:

RuntimeWarning: invalid value encountered in scalar divide return np.sum(all_loss) / ntokens
Epoch 1: Iter 1: Val loss nan, Val took 0.001s

Steps taken to diagnose:
1. Is it a learning rate issue? Lowered learning_rate in lora.py from 1.e-5 to 1.e-6
2. Does a lower or higher eps help in optimization? Passed eps=1.e-6 and then eps=1.e-5 to the Adam optimizer in lora.py
3. Were examples generated in ./generated/train.jsonl and ./generated/test.jsonl --> yes

**Expected behavior**
Training would proceed with numerical (not Nan) values for the validation loss

**Full error output**
ilab model train --model-dir ./merlinite-7b-base --input-dir ./generated --iters 5 --local
[INFO] Loading
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
dtype=mlx.core.float16
[INFO] Quantizing
Using model_type='mistral'
Loading pretrained model
Using model_type='mistral'
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Total parameters 1244.079M
Trainable parameters 1.704M
Loading datasets
Training
/Users/davidgold/Documents/local_python/instructlab/.venv/lib/python3.12/site-packages/instructlab/train/lora_mlx/lora.py:130: RuntimeWarning: invalid value encountered in scalar divide
  return np.sum(all_losses) / ntokens
Epoch 1: Iter 1: Val loss nan, Val took 0.001s

**Device Info (please complete the following information):**
 - Hardware Specs: Apple M3 Pro, 36GB
 - OS Version: macOS Sequoia 15.3.2
 - Python Version: Python 3.12.4
 - InstructLab Version (output of ilab sysinfo):
 -instructlab.version: 0.17.2
sys.version: 3.12.4 (main, Jul 25 2024, 22:11:22) [Clang 18.1.8 ]
sys.platform: darwin
os.name: posix
platform.release: 24.3.0
platform.machine: arm64
torch.version: 2.6.0
torch.backends.cpu.capability: NO AVX
torch.version.cuda: None
torch.version.hip: None
torch.cuda.available: False
torch.backends.cuda.is_built: False
torch.backends.mps.is_built: True
torch.backends.mps.is_available: True
llama_cpp_python.version: 0.2.75
llama_cpp_python.supports_gpu_offload: True

**Additional context**
references an internal repo stored in project directory e.g., instructlab/knowledge_repo

**Comment by dgoldwx2112** [comment]
### Resolution

I traced the NaN validation losses to the default Mistral + Accelerate MPS offload pipeline — FP16 quantization and placeholder tensors were underflowing. To fix this, I wrote a custom half-precision training loop that:

1. Loads the entire model in `torch.float16` directly onto MPS (no `meta` device leftovers).
2. Applies fixed-length padding and masks out pad tokens (so they don’t contribute to loss).
3. Unsqueezes inputs to produce proper 2D `attention_mask`s.
4. Wraps the model in a small LoRA adapter (`r=4`, `α=16`) and uses a tuned optimizer (`LR=5×10⁻⁶`, AdamW with warm-up scheduler).
5. Implements gradient clipping and early stopping.

This custom loop eliminated the NaNs and produced stable, finite train/validation losses, with early stopping kicking in at epoch 7.
